### PR TITLE
feature/toggle-switch

### PR DESCRIPTION
**Pull request by raphael-suter**
# Done is when (DoD):
- [x] Modifications within components are reflected by tests.
- [x] A self-review of the code has been performed.
- [x] The modified component properties have been added to the typescript typings.
- [x] Potential design changes have been reviewed by a designer.
- [x] The modifications are well documented (README.md, etc.) and showcased in the stories.
- [x] The modifications are documented in the code, particularly in hard-to-understand areas (focus on **WHY**).
- [x] The modifications are shortly added to CHANGELOG.md with the issue number.
- [x] The modifications still work in IE.
- [x] The modifications are according to our [Best Practices](https://github.com/axa-ch/patterns-library/blob/develop/CONTRIBUTION.md#best-practices)

### CODE REVIEW

File: src/components/10-atoms/toggle-switch/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    <script src="node_modules/@axa-ch/toggle-switch/dist/index.js"></script>
+  </body>
+</html>
+```
+
+## Properties
+
+### on
+
+The boolean `on` sets the on/off visual state.
```
Comment (markus-walther) [inline src/components/10-atoms/toggle-switch/README.md]: spelling: **B**oolean

Phrasing: either no 'The' or 'Boolean attribute/property'

File: src/components/10-atoms/toggle-switch/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  <body>
+    <axa-toggle-switch></axa-toggle-switch>
+    <script src="node_modules/@axa-ch/toggle-switch/dist/index.js"></script>
+  </body>
+</html>
+```
+
+## Properties
+
+### on
```
Comment (markus-walther) [inline src/components/10-atoms/toggle-switch/README.md]: **on** to my mind is not a great property name, but you may check with others for their opinion.

File: src/components/10-atoms/toggle-switch/index.scss [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      & + .a-toggle-switch__slider {
+        background-color: $color-malachite;
+      }
+
+      & + .a-toggle-switch__slider::before {
+        transform: translateX(20px);
+      }
+
+      &:disabled + .a-toggle-switch__slider {
+        background-color: #9FD9B4;
```
Comment (markus-walther) [inline src/components/10-atoms/toggle-switch/index.scss]: why not symbolic color?

File: src/components/10-atoms/toggle-switch/index.scss [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      }
+    }
+
+    &:disabled + .a-toggle-switch__slider {
+      background-color: #F5F5F5;
```
Comment (markus-walther) [inline src/components/10-atoms/toggle-switch/index.scss]: why not symbolic color?

File: src/components/10-atoms/toggle-switch/ui.test.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  await t.expect($axaElem.exists).ok();
+
+  const $axaElemShadow = await Selector(
+    () => document.querySelector(TAG).shadowRoot,
+    { dependencies: { TAG } }
+  );
+
+  const $axaElemShadowEl = await $axaElemShadow.find(CLASS);
+  await t.expect($axaElemShadowEl.exists).ok();
+});
```
Comment (markus-walther) [inline src/components/10-atoms/toggle-switch/ui.test.js:22]: There should also be a test for toggling.

File: src/components/10-atoms/toggle-switch/CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+## [Unreleased]
+
+- short description of bugfix/feature that is not released yet
```
Comment (markus-walther) [inline src/components/10-atoms/toggle-switch/CHANGELOG.md]: Make sure to remove this stuff before 1st release...

File: src/components/10-atoms/toggle-switch/README.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
+# AXA Toggle Switch
+
+Toggle Switches provide a UI element for toggling between two visual states, on and off.
```
Comment (markus-walther) [inline src/components/10-atoms/toggle-switch/README.md]: Phrasing: &lt;axa-toggle-switch&gt; provides....

File: src/components/10-atoms/toggle-switch/index.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+  static get styles() {
+    return css`
+      ${unsafeCSS(styles)}
+    `;
+  }
+
+  static get properties() {
+    return {
+      on: { type: Boolean },
```
Comment (markus-walther) [inline src/components/10-atoms/toggle-switch/index.js]: Would it make sense to reflect any of these props? Why or why not?

**Review by markus-walther — CHANGES_REQUESTED** [review]
Great start - I guess you are not fully finished, so take it as a quick in-between review.

Found a few things, nothing dramatic...

File: src/components/10-atoms/toggle-switch/README.md (hunk 2 above)
Comment (danyball) [reply src/components/10-atoms/toggle-switch/README.md]: Better rename that name to `active` or something.

File: src/components/10-atoms/toggle-switch/index.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        ?checked="${on}"
+        ?disabled="${disabled}"
+        @change=${onChange}
+      />
+    `;
+
+    return html`
+      <label class="a-toggle-switch">
+        ${inputElement}
+        <span class="a-toggle-switch__slider"></span>
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.js]: Why did you choose a span?

File: src/components/10-atoms/toggle-switch/index.scss [hunk 10]
```diff
@@ -0,0 +1,57 @@
+$toggle-switch-height: 30px;
+$toggle-switch-animation: 0.4s ease;
+
+:host {
+  display: inline-block;
+  height: $toggle-switch-height;
+}
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.scss:9]: Why do you need to style the :host ?

File: src/components/10-atoms/toggle-switch/index.scss [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    position: relative;
+    height: $toggle-switch-height;
+    width: 50px;
+    border-radius: 18px;
+    background-color: $color-light-grey;
+    cursor: pointer;
+    transition: $toggle-switch-animation;
+
+    &::before {
+      content: "";
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.scss:52]: Why do you need that?

File: src/components/10-atoms/toggle-switch/index.scss [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      position: absolute;
+      top: 2px;
+      left: 2px;
+      height: 26px;
+      width: 26px;
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.scss]: If you create variables for top and left you can calculate height and width here.

File: src/components/10-atoms/toggle-switch/package.json [hunk 13]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "@axa-ch/toggle-switch",
+  "version": "0.0.0-beta.0",
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/package.json:3]: Dont forget to set 1.0.0 before it gets merged to develop.

**Review by danyball — CHANGES_REQUESTED** [review]

File: src/components/10-atoms/toggle-switch/README.md [hunk 14]
```diff
@@ -55,13 +55,13 @@ Import the toggle-switch-defining script and use a toggle-switch like this:
 
 ## Properties
 
-### on
+### active
 
-The boolean `on` sets the on/off visual state.
+The Boolean attribute `active` sets the on/off visual state.
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/README.md]: It not just sets the `visual` state. Also the "real" state of the internal checkbox. So just remove the word "visual". Or just "....sets the state". but like u want....

File: src/components/10-atoms/toggle-switch/index.js [hunk 15]
```diff
@@ -47,6 +47,13 @@ class AXAToggleSwitch extends LitElement {
       </label>
     `;
   }
+
+  handleChange(event) {
+    this.active = event.target.checked;
+
+    // this.dispatchEvent(event);
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.js]: this comment is not explaining anything ;)

File: src/components/10-atoms/toggle-switch/index.js [hunk 16]
```diff
@@ -47,6 +47,13 @@ class AXAToggleSwitch extends LitElement {
       </label>
     `;
   }
+
+  handleChange(event) {
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.js:52]: Why did you add that handler? (i think i know.... but want to read your explanation)

File: src/components/10-atoms/toggle-switch/index.scss (hunk 11 above)
Comment (raphael-suter) [reply src/components/10-atoms/toggle-switch/index.scss:52]: > Why do you need that?

I need `content: ""` to create a pseudo-element.

File: src/components/10-atoms/toggle-switch/index.js (hunk 16 above)
Comment (raphael-suter) [reply src/components/10-atoms/toggle-switch/index.js:52]: > Why did you add that handler? (i think i know.... but want to read your explanation)

I added a handler to make sure that the `active` property is always updatet and reflected correctly.
& I think I should also dispatch an event and not only call the callback function, but I'll need some help with this.

Comment (danyball) [reply src/components/10-atoms/toggle-switch/index.js:52]: You dont need to dispatch an event. It is already dispatched by the browser when the `checked` value changed at the input element.

File: src/components/10-atoms/toggle-switch/index.react.d.ts [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+export interface AXAToggleSwitchProps {
+  on?: boolean;
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.react.d.ts]: change to active

File: src/components/10-atoms/toggle-switch/index.scss [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+$toggle-switch-spacing: 2px;
+
+.a-toggle-switch {
+  display: block;
+  position: relative;
+  height: $toggle-switch-height;
+  width: $toggle-switch-width;
+
+  &__input {
+    display: none;
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.scss]: Please ask @markus-walther why he dont just set `display: none` at the `checkbox`component. His code there is: 
```
.a-checkbox__input {
    position: absolute;
    width: 1px;
    height: 1px;
    clip: rect(0, 0, 0, 0);
    margin: -1px;
    overflow: hidden;
  } 
```

File: src/components/10-atoms/toggle-switch/index.scss [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+    &:checked {
+      & + .a-toggle-switch__slider {
+        background-color: $color-malachite;
+      }
+
+      & + .a-toggle-switch__slider::before {
+        $distance: $toggle-switch-width - $toggle-switch-height;
+        transform: translateX($distance);
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.scss]: Move `$distance` to the other variables on top of file. Or remove that variable and calculate directly: `translateX(w - h)`

File: src/components/10-atoms/toggle-switch/index.scss [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+    right: 0;
+    bottom: 0;
+    left: 0;
+    border-radius: 18px;
+    background-color: $color-light-grey;
+    transition: $toggle-switch-animation;
+    cursor: pointer;
+
+    &::before {
+      $size: $toggle-switch-height - 2 * $toggle-switch-spacing;
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.scss]: Move $size to the other variables on top of file. (So all the variables can be seen with one view)

File: src/components/10-atoms/toggle-switch/index.scss [hunk 21]
```diff
@@ -0,0 +1,60 @@
+$toggle-switch-animation: 0.4s ease;
+$toggle-switch-height: 30px;
+$toggle-switch-width: 50px;
+$toggle-switch-spacing: 2px;
+
+.a-toggle-switch {
+  display: block;
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.scss:12]: With `block` its not possible to place other elements next to the switch. But this could be a possible workflow. Why not `inline-block`?

**Review by danyball — CHANGES_REQUESTED** [review]

File: src/components/10-atoms/toggle-switch/index.scss [hunk 22]
```diff
@@ -11,7 +11,7 @@ $toggle-switch-circle-size: $toggle-switch-height - 2 * $toggle-switch-spacing;
   width: $toggle-switch-width;
 
   &__input {
-    display: none;
+    opacity: 0;
```
Comment (danyball) [inline src/components/10-atoms/toggle-switch/index.scss:19]: Please add a short comment why you choose that variant.

File: src/components/10-atoms/toggle-switch/index.scss (hunk 21 above)
Comment (danyball) [reply src/components/10-atoms/toggle-switch/index.scss:12]: Analog to our axa-button you can write:
```
:host {
  display: inline-block;
}
.a-toggle-switch {
  display: block;
```

**Review by danyball — CHANGES_REQUESTED** [review]

**Review by danyball — DISMISSED** [review]

**Review by danyball — APPROVED** [review]
